daffodil-tdml-lib/src/main/scala/org/apache/daffodil/tdml/RunnerFactory.scala

```diff
@@ -21,69 +21,94 @@ import org.apache.daffodil.util.Misc
-  // roundTrip - on test case
-  // defaultRoundTrip - on test suite
-  // defaultRoundTripDefault - on runner aka test suite factory
-  // defaultRoundTripDefaultDefault - on runner factory
-  //
   def defaultRoundTripDefaultDefault: RoundTrip = NoRoundTrip
-  def defaultValidationDefaultDefault = "off"
+  def defaultValidationDefaultDefault: String = "off"
 
   /**
    * Default for what DFDL implementations to run tests against.
    *
    * A test or test suite can override this to specify more or different implementations
    * that the test should pass for.
    */
-  def defaultImplementationsDefaultDefault = Seq("daffodil", "ibm")
+  def defaultImplementationsDefaultDefault: Seq[String] = Seq("daffodil", "ibm")
 
   /**
    * By default we don't run Daffodil negative TDML tests against cross-testers.
    * The error messages are simply too varied.
    *
    * Negative tests must fail, but error messages aren't compared.
    */
-  def defaultShouldDoErrorComparisonOnCrossTests = false
+   def defaultShouldDoErrorComparisonOnCrossTests: Boolean = false
 
   /**
    * By default we don't cross test warning messages because they are too varied.
    */
-  def defaultShouldDoWarningComparisonOnCrossTests = false
+  val defaultShouldDoWarningComparisonOnCrossTests: Boolean = false
+}
 
+class RunnerOpts(
+  var validateTDMLFile: Boolean = true,
+  var validateDFDLSchemas: Boolean = true,
+  var compileAllTopLevel: Boolean = false,
+  var defaultRoundTripDefault: RoundTrip,
+  var defaultValidationDefault: String,
+  var defaultImplementationsDefault: Seq[String],
+  var shouldDoErrorComparisonOnCrossTests: Boolean,
+  var shouldDoWarningComparisonOnCrossTests: Boolean) {
+  // Yes, that's a lot of defaults.....
+  // but really it is 3-tiers deep:
+  // roundTrip - on test case
+  // defaultRoundTrip - on test suite
+  // defaultRoundTripDefault - on runner aka test suite factory
+  // defaultRoundTripDefaultDefault - on runner factory
+}
+
+/**
+ * Creates the DFDLTestSuite object lazily, so the file isn't read into memory
+ * and parsed unless you actually try to run a test using it.
+ *
+ * Creates the DFDLTestSuite only once.
+ *
+ * Provides a reset method to be called from @AfterClass to drop
+ * the test suite object (and avoid memory leak).
+ *
+ * Note: I have verified that this does get called after each test suite has been run.
+ *
+ * defaultRoundTripDefault if true the round trip default for the test suite will be
+ * this value, if the test suite does not specify defaultRoundTrip attribute.
+ *
+ * defaultRoundTripDefaultDefault
+ */
+object Runner {
+  def apply(dir: String, file: String): Runner =
+    new Runner(dir, file, RunnerOpts())
+
+  def apply(dir: String, file: String, options: RunnerOpts): Runner =
+    new Runner(dir, file, options)
+
+  def apply(elem: scala.xml.Elem): Runner =
+    new Runner(elem, RunnerOpts())
+
+  def apply(elem: scala.xml.Elem, options: RunnerOpts): Runner =
+    new Runner(elem, options)
 }
```

Comment:
I'm concerned that this could break existing Runners not in the daffodil repo. Other schema repos use this ``Runner`` for testing and they could currently provide options which would no longer compile once this change is made. I'm wondering if it makes sense to keep an apply the old dir/file apply method with all options defaulted and creates the new ``RunnerOpts`` using those values? We can deprecate that ``apply`` method so other repos will still work but get a message to fix them?

Reply:
I would second this. We need the original Runner factory method, as there are dozens of DFDL schemas we know of, and possibly others we don't know of or have access to, which are using the existing Runner factory.

Reply:
In commit b123d9f0ed174df22f87b564dbe38cab853e8362 I restored the old apply methods.